### DISCUSSION TITLE
The script works for only visible nodes, help sought to make it work for all nodes

### DISCUSSION THREAD

**Discussion by biswajit-goswami** [freeplane/freeplane#1144] [Help] [answered]
Hi,
I have the following script that attempts to filter all the nodes which have an attribute as the text of the selected node. THe problem is that it filters only the visible nodes, not all the nodes in the map. can you plz have a look and point to what changes may be needed to make it work for all nodes of the map:

def selectedNodes=c.selecteds
def showAncestors = false
def showDescendants = true
def nodeList = []
if( selectedNodes.size() ==1 ) {
    attributeName=selectedNodes[0].text
	c.findAll().each	
	{
	def nodeAttributes = it.attributes 
	nodeAttributes.names.each { name ->
			if (name != null && name==attributeName)
			{ 
				// Select the node 
				nodeList.add(it)
			}
		}
	}

// Apply filter

map.filter(showAncestors, showDescendants, {it in nodeList})

}

**Comment by macmarrum** [comment]
What makes your nodes not visible?
- Are they folded?
- Are they hidden?
- Are they outside of a jump-in?
- Are they filtered by another filter?

It would be helpful if you could attach a sample mindmap illustrating your case with a step-by-step on how to get to your problem.

**Reply by biswajit-goswami** [reply]
By not visible, I meant folded nodes. If I unfold the nodes, then the filter works on the nodes which got unfolded.

In the attached map, I try to filter the map with an attribute name 'b' ...there is a node 'd' which has that attribute. But if I keep it folded, the filter ignores it. I have to unfold that node.

![image](https://github.com/freeplane/freeplane/assets/11291623/135eef92-7ff0-40fc-9251-4e25624cd3a3)

[TestMap.zip](https://github.com/freeplane/freeplane/files/11483286/TestMap.zip)

**Reply by abc16361** [reply]
I added lines to select those nodes with the attribute, the folded nodes will be unfolded.
```
def nodeAttrs = c.find{it.attributes.getNames().contains(attributeName)}
c.select(nodeAttrs)
```

**Reply by biswajit-goswami** [reply]
Thanks, it selects the right nodes, including folded ones. But when I try to filter then out using the below code:

def nodeAttrs = c.find{it.attributes.getNames().contains(attributeName)}
map.filter(showAncestors, showDescendants, {it in nodeAttrs})

it again filters the unfolded nodes..any hints?

**Reply by macmarrum** [reply, marked as answer]
@biswajit-goswami 
When a node is folded, it stays folded whether filter is applied or not.
One way to make nodes visible is to select them, as suggested by abc16361.

Here's your script with an addition of `c.select(nodeList)` which should do the trick.
I believe it has no relation to `c.find()` vs `c.findAll()`

```groovy
def selectedNodes = c.selecteds
def showAncestors = false
def showDescendants = true
def nodeList = []
if (selectedNodes.size() == 1) {
    attributeName = selectedNodes[0].text
    c.findAll().each {
        def nodeAttributes = it.attributes
        nodeAttributes.names.each { name ->
            if (name != null && name == attributeName) {
                // Select the node
                nodeList.add(it)
            }
        }
    }

    // Apply filter
    map.filter(showAncestors, showDescendants, { it in nodeList })

    // select nodes, so that they are made visible
    c.select(nodeList)
}
```

Here's another way to achieve what you described, which calls `Filter->Select nodes matching filter` at the end:

```groovy
def attributeName = node.text
def showAncestors = false
def showDescendants = true
node.mindMap.filter(showAncestors, showDescendants, { it.attributes.containsKey(attributeName) })
menuUtils.executeMenuItems(['SelectFilteredNodesAction'])
```

**Reply by biswajit-goswami** [reply]
Thanks @macmarrum , @abc16361
